File: src/my_py_pkg/my_py_pkg/number_counter.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from example_interfaces.msg import Int64
from example_interfaces.srv import SetBool

class Number_Counter(Node): #create the class and inharet from node
    def __init__(self): 
        super().__init__("Number_Counter") #call the init from the Node class
        self.publiser_=self.create_publisher(Int64,"number_counter",10)
        self.subscriber_=self.create_subscription(Int64,"number",self.callback_subscriber_number,10)
        self.get_logger().info("number_counter is running...")
        self.counter_=0
        self.server_= self.create_service(SetBool,"rest_counter",self.callback_service_setbool)
     #here we used the same subcriber function to publish and this is a comman way to publish while you sub from another topic
    
    def callback_subscriber_number(self, mesg: Int64):
        self.counter_+=mesg.data
        #self.get_logger().info(str(mesg.data))
        masage = Int64()
        masage.data=self.counter_
        self.publiser_.publish(masage)
    
    def callback_service_setbool(self, request: SetBool.Request , responce: SetBool.Response ):
        if request.data:
            self.counter_= 0
            responce._success= True
            responce.message= "the counter is int to zero sucssfully"
            self.get_logger().info(f"{responce.message} : {responce._success} ")
        else:
            responce._success= False
            responce._message= "Watting to call true ..."
            self.get_logger().warn(f"{responce.message} : {responce._success} ")

        return responce
        
        



def main(args=None):
    rclpy.init(args=args)   
    node=Number_Counter()   
    rclpy.spin(node)        
    rclpy.shutdown()        



#if the name main called in the terminal call the main function
if __name__ == "__main__":
    main()

File: number_counter.cpp
#include "rclcpp/rclcpp.hpp" //importing the ros2 lib if it shows a error open vs from the terminal
#include "example_interfaces/msg/int64.hpp"
#include "example_interfaces/srv/set_bool.hpp"
using namespace std::chrono_literals;
using namespace std::placeholders;

class Number_Counter : public rclcpp::Node //create a class and inharet from the rclcpp
{
    public: //puplic
        Number_Counter() : Node("Number_Counter") , counter_(0)
        {
            //assign the v and the methods
            this->publisher_=this->create_publisher<example_interfaces::msg::Int64>("number_counter",10);
            this->subscriber_=this->create_subscription<example_interfaces::msg::Int64>("CPP_number",10,
            std::bind(&Number_Counter::callback_subscriber,this,_1));
            RCLCPP_INFO(this->get_logger(),"The counter started...");


            this->server_= this->create_service<example_interfaces::srv::SetBool>("number_reset_counter",
                                                std::bind(&Number_Counter::callback_setbool,this, _1, _2));

        }   
    private:
    //define the v of the class and the private methods
    rclcpp::Publisher<example_interfaces::msg::Int64>::SharedPtr publisher_;
    rclcpp::Subscription<example_interfaces::msg::Int64>::SharedPtr subscriber_;
    rclcpp::Service<example_interfaces::srv::SetBool>::SharedPtr server_;
    int64_t counter_ ;

    void callback_subscriber(const example_interfaces::msg::Int64::SharedPtr msg)
    {
        auto temp = example_interfaces::msg::Int64();
        this->counter_+= msg->data;
        temp.data= counter_;

        this->publisher_->publish(temp);
    }

    void callback_setbool(const example_interfaces::srv::SetBool::Request::SharedPtr request ,
                          const example_interfaces::srv::SetBool::Response::SharedPtr responce)
    {
        if(request->data)
        {
            this->counter_=0;
            responce->success= 1;
            responce->message= "The counter is zero now: %d",responce->success ;
            RCLCPP_INFO(this->get_logger(),"The counter is zero now: %d",responce->success);
        }
        else
        {
            responce->success= 0;
            responce->message= "Sending false Waite for  true : %d",responce->success ;
            RCLCPP_WARN(this->get_logger(),"Sending false Waite for  true : %d",responce->success);
        }
    }
};

int main(int argc,char **argv)
{
    rclcpp::init(argc,argv);
    auto node = std::make_shared<Number_Counter>(); //create the shared pointer of the my node
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;

}
